src/coreclr/nativeaot/Runtime/EHHelpers.cpp

```diff
@@ -536,6 +536,57 @@ int32_t __stdcall RhpVectoredExceptionHandler(PEXCEPTION_POINTERS pExPtrs)
         return EXCEPTION_CONTINUE_SEARCH;
     }
 
+    // the following would work on ARM64 as well, but there is no way to test right now.
+#ifdef TARGET_AMD64
+
+#ifndef STATUS_RETURN_ADDRESS_HIJACK_ATTEMPT
+#define STATUS_RETURN_ADDRESS_HIJACK_ATTEMPT ((uintptr_t)0x80000033L)
+#endif
+
+    if (faultCode == STATUS_RETURN_ADDRESS_HIJACK_ATTEMPT)
+    {
+        Thread * pThread = ThreadStore::GetCurrentThreadIfAvailable();
+        if (pThread == NULL || !pThread->IsCurrentThreadInCooperativeMode())
+        {
+            // if we are not in coop mode, this cannot be our hijack
+            // Perhaps some other runtime is responsible.
+            return EXCEPTION_CONTINUE_SEARCH;
+        }
+
+        // Sanity check. The thread should be hijacked by us.
+        _ASSERTE_ALL_BUILDS(pThread->IsHijacked());
+
+        PCONTEXT interruptedContext = pExPtrs->ContextRecord;
+        bool areShadowStacksEnabled = PalAreShadowStacksEnabled();
+        if (areShadowStacksEnabled)
+        {
+            // OS should have fixed the SP value to the same as we`ve stashed for the hijacked thread
+            _ASSERTE(*(size_t *)interruptedContext->GetSp() == (uintptr_t)pThread->GetHijackedReturnAddress());
+
+            // When the CET is enabled, the interruption happens on the ret instruction in the calee.
+            // We need to "pop" rsp to the caller, as if the ret has consumed it.
+            interruptedContext->SetSp(interruptedContext->GetSp() + 8);
+        }
+
+        // Change the IP to be at the original return site, as if we have returned to the caller.
+        // That IP is an interruptible safe point, so we can suspend right there.
+        uintptr_t origIp = interruptedContext->GetIp();
```

Comment:
@jakobbotsch this style of hijacking can hit some asserts if the caller is missing safepoint for the call site. (i.e. MinOpt). The code will still work, but asserts will need to be adjusted.

Re: https://github.com/dotnet/runtime/pull/103950

Reply:
Does this work the same way in regular CoreCLR?

Reply:
It could. I think all hijacks could work this way and it will be simpler. 

I.E no need for special frames, interrogating the containing method info for kinds of GC returns, capturing that somewhere,…

Reply:
That is assuming we bumped r2r version

Reply:
> It could.

It is important that it does. Our native AOT testing strategy depends on minimizing differences between regular CoreCLR and native AOT.

Reply:
Looks like r2r version was incremented. 👍

Reply:
Since CET compat will be the default, this code path will be exercised on everything win10 and newer, so this will be reasonably tested. 

I can move CoreCLR on the same plan, so we have more confidence. I was planning to do that anyways.

Reply:
Here a PR that uses this pattern in CoreCLR CET hijacks and in GC stress. 
https://github.com/dotnet/runtime/pull/104198